packages/driver/src/multidomain/communicator.ts

```diff
@@ -0,0 +1,110 @@
+import { EventEmitter } from 'events'
+
+const CROSS_DOMAIN_PREFIX = 'cross:domain:'
+
+/**
+ * Primary domain communicator. Responsible for sending/receiving events throughout
+ * the driver responsible for multi-domain communication, as well as sending/receiving events to/from the
+ * spec bridge communicator, respectively.
+ *
+ * The 'postMessage' method is used to send events to the spec bridge communicator, while
+ * the 'message' event is used to receive messages from the spec bridge communicator.
+ * All events communicating across domains are prefixed with 'cross:domain:' under the hood.
+ * See https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage for more details.
+ * @extends EventEmitter
+ */
+export class PrimaryDomainCommunicator extends EventEmitter {
+  private windowReference
+  private crossDomainDriverWindow
+
+  /**
+   * Initializes the event handler to receive messages from the spec bridge.
+   * @param {Window} win - a reference to the window object in the primary domain.
+   * @returns {Void}
+   */
+  initialize (win) {
+    if (this.windowReference) return
+
+    this.windowReference = win
+
+    this.windowReference.top.addEventListener('message', ({ data, source }) => {
+      // currently used for tests, can be removed later
+      if (data?.actual) return
+
+      // check if message is cross domain and if so, feed the message into
+      // the cross domain bus with args and strip prefix
+      if (data?.event?.includes(CROSS_DOMAIN_PREFIX)) {
+        const messageName = data.event.replace(CROSS_DOMAIN_PREFIX, '')
+
+        // NOTE: need a special case here for 'window:before:load'
+        // where we need to set the crossDomainDriverWindow to source to
+        // communicate back to the iframe
+        if (messageName === 'window:before:load') {
+          this.crossDomainDriverWindow = source
+        } else {
+          this.emit(messageName, data.data)
+        }
+
+        return
+      }
+
+      // TODO: how do we want to log unexpected messages, if at all?
+      // eslint-disable-next-line no-console
+      console.log('Unexpected postMessage:', data)
```

Comment:
Perhaps like this using `debug` https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cy/commands/navigation.ts#L14-L15